avocado/plugins/runner_nrunner.py

```diff
@@ -103,23 +105,24 @@ def run_suite(self, job, test_suite):
                                                     job.result,
                                                     early_state)
 
-            statuses = []
             task.status_services = []
             for status in task.run():
+                status_repo.process_message(status)
                 result_dispatcher.map_method('test_progress', False)
-                statuses.append(status)
+
                 if status['status'] not in ["started", "running"]:
                     break
 
             # test execution time is currently missing
             # since 358e800e81 all runners all produce the result in a key called
             # 'result', instead of 'status'.  But the Avocado result plugins rely
             # on the current runner approach
-            test_state = {'status': statuses[-1]['result'].upper()}
+            this_task_data = status_repo.get_task_data(task.identifier)
+            test_state = {'status': this_task_data[-1]['result'].upper()}
```

Comment:
> I was expecting something like `task.status` or `tasks.get_status()`. I understand that this would imply in change a lot here. But from a usability/maintenance perspective, this code is a little bit strange to me and I would try a different approach.
> 
> Here, I see Avocado as a pipeline where we have a few main objects (job, suites, and tasks) that are moving through this pipeline and get updated over time. So I would try, **whenever possible**, to move the objects and update them instead of moving raw data. So, why are you storing "raw data" inside the `status_repo` instead of the `task` objects is something that I'm trying to understand here.
> 

There will be the handling of the "pipeline" indeed, but this is a more limited class that is just a local repository of the "history" (may even be partial) of a Task's status.  The reason for keeping this data include:

* "what was the last time a task sent a 'running' message?" as part of a timeout handling
* which tasks have reported a "result"?

So it's just a data repository indeed.  The logic will be built on this data.

> Maybe we need some entity that knows all the tasks that are running and would return the task object based on an ID. So it could be easy to manipulate think and avoid things like `get_task_data`. (We probably will need some "SerDes" methods inside the task class to help here). `task.as_json()` and `Task.from_json()` for instance.
> 
> So, yes, I understand that we are running out of time, and maybe we could discuss this after LTS, but at a first look, this status repo could store the task objects and be responsible for updating their status based on the messages that receive.
> 

Adding status data into the `Task` instances themselves IMO has advantages but also drawbacks... hard to go through all of them at this moment/point though.

> ** We don't need to discuss this right now **

I'm certainly inspired by the suggestions/ideas here, and hopefully the reason will become apparent very soon.

Reply:
I will keep my energy for future discussions on this. :)